File: tests/test_agarwal.cpp
/*
 * Copyright (C) 2018-2020 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */


#include <array>
#include <catch2/catch.hpp>
#include <map>

// Defines the geometry kernel
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/xml_parser.hpp>

#include "helpers/CustomCatchTemplate.h"
#include "movetk/Simplification.h"
#include "movetk/metric/Distances.h"
#include "movetk/metric/Norm.h"

// Add a message to a Catch2 require
#define REQUIRE_MESSAGE(cond, msg) \
	do {                             \
		INFO(msg);                     \
		REQUIRE(cond);                 \
	} while ((void)0, 0)

struct AgarwalTestcase {
	std::string polyline_input;
	// Ipe input for single edge that will define the epsilon for the testcase
	std::string epsilon_line_input;
	std::vector<std::size_t> expectedInds;

	static constexpr const char* test_data_file = "data/agarwal_simplification_tests.xml";

	static std::map<std::string, AgarwalTestcase> load_test_cases() {
		std::map<std::string, AgarwalTestcase> test_cases;
		boost::property_tree::ptree tree;
		boost::property_tree::read_xml(test_data_file, tree);
		const auto& tests_tree = tree.get_child("tests");
		for (const auto& node : tests_tree) {
			const auto test_name = node.second.get<std::string>("name");
			AgarwalTestcase test_case;
			test_case.polyline_input = node.second.get<std::string>("input.ipeselection.path");
			test_case.epsilon_line_input = node.second.get<std::string>("expected_line.ipeselection.path");
			const auto expected_indices_string = node.second.get<std::string>("expected_indices");
			test_helpers::vector_from_string(expected_indices_string, test_case.expectedInds);
			test_cases[test_name] = test_case;
		}
		return test_cases;
	}
};

TEMPLATE_LIST_TEST_CASE("Check that the simplifications are correct",
                        "[agarwal_simplification][simplification]",
                        movetk::test::AvailableBackends) {
	using MovetkGeometryKernel = typename TestType::MovetkGeometryKernel;
	// The norm to be used in weak Frechet distance computations.
	using Norm = movetk::metric::FiniteNorm<MovetkGeometryKernel, 2>;
	using NT = typename MovetkGeometryKernel::NT;
	using SFR = movetk::metric::StrongFrechet<MovetkGeometryKernel,
	                                          movetk::metric::squared_distance_d<MovetkGeometryKernel, Norm>>;
	using SqDistance = movetk::metric::squared_distance_d<MovetkGeometryKernel, Norm>;
	movetk::simplification::Agarwal<MovetkGeometryKernel, SqDistance> simplifier;
	simplifier.setTolerance(0.0001);
	using PointList = std::vector<typename MovetkGeometryKernel::MovetkPoint>;
	const auto parseIpe = [](const std::string& string_data, PointList& output) {
		test_helpers::parse_ipe_path_contents<MovetkGeometryKernel>(string_data, output);
	};
	const auto test_cases = AgarwalTestcase::load_test_cases();
	for (const auto& [test_case_name, test_case] : test_cases) {
		SECTION(test_case_name) {
			std::cout << test_case_name << '\n';
			PointList points, epsilonPath;

			parseIpe(test_case.polyline_input, points);
			parseIpe(test_case.epsilon_line_input, epsilonPath);

			// Requested epsilon
			const auto epsilon = std::sqrt(SqDistance()(epsilonPath[0], epsilonPath[1]));
			simplifier.setEpsilon(epsilon);

			std::vector<decltype(points.begin())> output;

			simplifier(points.begin(), points.end(), std::back_inserter(output));
			REQUIRE(output.size() == test_case.expectedInds.size());
			// Compute indices of iterators
			std::vector<std::size_t> inds;
			for (const auto& el : output) {
				inds.push_back(std::distance(points.begin(), el));
			}
			REQUIRE(inds == test_case.expectedInds);
		}
	}
	SECTION("Single point and empty polyline") {
		std::vector<typename MovetkGeometryKernel::MovetkPoint> points;
		points.push_back(movetk::geom::MakePoint<MovetkGeometryKernel>()({(NT)2.0, (NT)5000.}));

		simplifier.setEpsilon(1.0);

		std::vector<decltype(points.begin())> output;

		simplifier(points.begin(), points.end(), std::back_inserter(output));

		REQUIRE(output.size() == 1);
		REQUIRE(output[0] == points.begin());

		// Now with empty
		output.clear();
		points.clear();

		simplifier(points.begin(), points.end(), std::back_inserter(output));
		REQUIRE(output.size() == 0);
	}
}

File: py/test/TestAgarwal.py
#
# Copyright (C) 2018-2020 HERE Europe B.V.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
# License-Filename: LICENSE
#
import unittest

from TestHelpers import parseIpePath, points_approximately_equal

from PyMoveTk.simplification import agarwal

class AgarwalTest:
    def __init__(self, polyline, epsilonLine, expectedComplexity=None, expectedIndices=None):
        self.polyline = polyline
        self.epsilonLine = epsilonLine
        self.expectedComplexity = expectedComplexity
        self.expectedIndices = expectedIndices
    def test(self, testCaseCls):
        # Requested epsilon
        epsilon = (self.epsilonLine[0] - self.epsilonLine[1]).norm()
        simplification = agarwal(self.polyline, epsilon)
        # Assert
        testCaseCls.assertEqual(len(simplification),self.expectedComplexity)
        # Verify the points at the indices are selected
        for i in range(len(simplification)):
            # Soft compare
            testCaseCls.assertTrue(points_approximately_equal(simplification[i], self.polyline[self.expectedIndices[i]]))

class TestAgarwal(unittest.TestCase):
    def test_spike_simplification_correct(self):
        polyline = parseIpePath("""<ipeselection pos="224 480">
            <path stroke="0">
            96 448 m
            128 448 l
            144 448 l
            192 448 l
            208 512 l
            224 448 l
            336 448 l
            352 448 l
            384 448 l
            </path>
            </ipeselection>""")
        epsilonPath = parseIpePath("""<ipeselection pos = "160 448">
            < path stroke = "black">
            160 448 m
            160 464 l
            </path>
            </ipeselection>
        )""")
        tester = AgarwalTest(polyline, epsilonPath, expectedComplexity=5, expectedIndices=[0,3,4,5,8])
        tester.test(self)
    
    def test_spike_simplification_larger_eps_correct(self):
        polyline = parseIpePath("""<ipeselection pos="224 480">
            <path stroke="0">
            96 448 m
            128 448 l
            144 448 l
            192 448 l
            208 512 l
            224 448 l
            336 448 l
            352 448 l
            384 448 l
            </path>
            </ipeselection>""")
        epsilonPath = parseIpePath("""<ipeselection pos = "160 448">
            < path stroke = "black">
            192 448 m
            168.371 489.355 l
            </path>
            </ipeselection>
        )""")
        tester = AgarwalTest(polyline, epsilonPath, expectedComplexity=4, expectedIndices=[0,4,5,8])
        tester.test(self)
    def test_spike_simplification_largest_eps_correct(self):
        polyline = parseIpePath("""<ipeselection pos="224 480">
            <path stroke="0">
            96 448 m
            128 448 l
            144 448 l
            192 448 l
            208 512 l
            224 448 l
            336 448 l
            352 448 l
            384 448 l
            </path>
            </ipeselection>""")
        epsilonPath = parseIpePath("""<ipeselection pos = "160 448">
            < path stroke = "black">
            208 448 m
            208 512 l
            </path>
            </ipeselection>
        )""")
        tester = AgarwalTest(polyline, epsilonPath, expectedComplexity=2, expectedIndices=[0,8])
        tester.test(self)

#     SECTION("Single segment")
#     {
#         std::vector<MovetkGeometryKernel::MovetkPoint> points;

#         test_helpers::parseIpePath(R"IPE(<ipeselection pos="224 480">
#             <path stroke="0">
#             96 448 m
#             128 448 l
#             </path>
#             </ipeselection>
#         )IPE", points);
#         std::vector<MovetkGeometryKernel::MovetkPoint> epsilonPath;
#         test_helpers::parseIpePath(R"IPE(
#             <ipeselection pos="208 464">
#             <path stroke="black">
#             0 0 m
#             0 5 l
#             </path>
#             </ipeselection>
#         )IPE", epsilonPath);

#         // Requested epsilon
#         const auto epsilon = std::sqrt(SqDistance()(epsilonPath[0], epsilonPath[1]));
#         simplifier.setEpsilon(epsilon);

#         std::vector<decltype(points.begin())> output;

#         simplifier(points.begin(), points.end(), std::back_inserter(output));
#         REQUIRE(output.size() == 2);
#         // Compute indices of iteratorss
#         std::vector<std::size_t> inds;
#         for (const auto& el : output)
#         {
#             inds.push_back(std::distance(points.begin(), el));
#         }
#         std::vector<std::size_t> expectedInds = { 0, 1 };
#         REQUIRE(inds == expectedInds);
#     }
#     SECTION("Single point and empty polyline")
#     {
#         std::vector<MovetkGeometryKernel::MovetkPoint> points;
#         points.push_back(movetk::geom::MakePoint<MovetkGeometryKernel>()({ (NT)2.0, (NT)5000. }));

#         simplifier.setEpsilon(1.0);

#         std::vector<decltype(points.begin())> output;

#         simplifier(points.begin(), points.end(), std::back_inserter(output));

#         REQUIRE(output.size() == 1);
#         REQUIRE(output[0] == points.begin());

#         // Now with empty
#         output.clear();
#         points.clear();

#         simplifier(points.begin(), points.end(), std::back_inserter(output));
#         REQUIRE(output.size() == 0);
#     }
# }